Mathlib/Data/Vector.lean

```diff
@@ -0,0 +1,257 @@
+/-
+Copyright (c) 2016 Microsoft Corporation. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Leonardo de Moura
+-/
+import Mathlib.Mathport.Rename
+import Std.Data.List.Basic
+import Std.Data.List.Lemmas
+import Mathlib.Init.Data.List.Basic
+import Mathlib.Init.Data.List.Lemmas
+
+/-!
+The type `Vector` represents lists with fixed length.
+-/
+
+universe u v w
+/-- `Vector α n` is the type of lists of length `n` with elements of type `α`. -/
+def Vector (α : Type u) (n : ℕ) :=
+  { l : List α // l.length = n }
+#align vector Vector
+
+namespace Vector
+
+variable {α : Type u} {β : Type v} {φ : Type w}
+
+variable {n : ℕ}
+
+instance [DecidableEq α] : DecidableEq (Vector α n) := by
+  unfold Vector
+  infer_instance
+
+/-- the empty vector with elements of type `α` -/
+@[match_pattern]
+def nil : Vector α 0 :=
+  ⟨[], rfl⟩
+#align vector.nil Vector.nil
+
+/-- If `a : α` and `l : Vector α n`, then `cons a l`, is the vector of length `n + 1`
+whose first element is a and with l as the rest of the list. -/
+@[match_pattern]
+def cons : α → Vector α n → Vector α (Nat.succ n)
+  | a, ⟨v, h⟩ => ⟨a :: v, congrArg Nat.succ h⟩
+#align vector.cons Vector.cons
+
+
+/-- The length of a vector -/
+@[reducible, nolint unusedArguments]
+def length (_ : Vector α n) : ℕ :=
+  n
+#align vector.length Vector.length
+
+open Nat
+
+/-- the first element of a vector with length at least `1` -/
+def head : Vector α (Nat.succ n) → α
+  | ⟨[], h⟩ => by contradiction
+  | ⟨a :: _, _⟩ => a
+#align vector.head Vector.head
+
+/-- the head of a vector obtained by prepending is the element prepended -/
+theorem head_cons (a : α) : ∀ v : Vector α n, head (cons a v) = a
+  | ⟨_, _⟩ => rfl
+#align vector.head_cons Vector.head_cons
+
+/-- the tail of a vector, with an empty vector having empty tail  -/
+def tail : Vector α n → Vector α (n - 1)
+  | ⟨[], h⟩ => ⟨[], congrArg pred h⟩
+  | ⟨_ :: v, h⟩ => ⟨v, congrArg pred h⟩
+#align vector.tail Vector.tail
+
+/-- the tail of a vector obtained by prepending is the vector prepended to -/
```

Comment:
Note that theorems don't need docstrings although they are most welcome! It's definitions which we demand docstrings for in mathlib4.